optuna/integration/dask.py

```diff
@@ -0,0 +1,738 @@
+            study_id=study_id,
+            deepcopy=deepcopy,
+        )
+        return [_serialize_frozentrial(t) for t in trials]
+
+    def get_n_trials(
+        self,
+        comm: "distributed.comm.tcp.TCP",
+        storage_name: str,
+        study_id: int,
+        state: Optional[TrialState] = None,
+    ) -> int:
+        return self.get_storage(storage_name).get_n_trials(
+            study_id=study_id,
+            state=state,
+        )
+
+    def read_trials_from_remote_storage(
+        self,
+        comm: "distributed.comm.tcp.TCP",
+        storage_name: str,
+        study_id: int,
+    ) -> None:
+        return self.get_storage(storage_name).read_trials_from_remote_storage(study_id=study_id)
+
+
+def _register_with_scheduler(
+    dask_scheduler: "distributed.Scheduler", storage: str, name: str
+) -> None:
+    if "optuna" not in dask_scheduler.extensions:
+        ext = _OptunaSchedulerExtension(dask_scheduler)
+    else:
+        ext = dask_scheduler.extensions["optuna"]
+
+    if name not in ext.storages:
+        ext.storages[name] = optuna.storages.get_storage(storage)
+
+
+def _use_basestorage_doc(func: Callable) -> Callable:
+    method = getattr(optuna.storages.BaseStorage, func.__name__, None)
+    if method is not None:
+        # Ensure BaseStorage and DaskStorage have the same signature
+        assert inspect.signature(func) == inspect.signature(method)
+        # Overwrite docstring if one does not exist already
+        if func.__doc__ is not None:
+            func.__doc__ = method.__doc__
+    return func
+
+
+@experimental("2.4.0")
+class DaskStorage(optuna.storages.BaseStorage):
+    """Dask-compatible storage class.
+
+    This storage class wraps a Optuna storage class (e.g. Optuna’s in-memory or sqlite storage)
+    and is used to run optimization trials in parallel on a Dask cluster.
+    The underlying Optuna storage object lives on the cluster’s scheduler and any method calls on
+    the ``DaskStorage`` instance results in the same method being called on the underlying Optuna
+    storage object.
+
+    See `this example <https://github.com/optuna/optuna/blob/master/
+    examples/dask_simple.py>`__
+    for how to use ``DaskStorage`` to extend Optuna's in-memory storage class to run across
+    multiple processes.
+
+    Args:
+        storage:
+            Optuna storage url to use for underlying Optuna storage class to wrap
+            (e.g. ``None`` for in-memory storage, ``sqlite:///example.db`` for SQLite storage).
+            Defaults to ``None``.
+
+        name:
+            Unique identifier for the Dask storage class. If not provided, a random name
+            will be generated.
+
+        client:
+            Dask ``Client`` to connect to. If not provided, will attempt to find an
+            existing ``Client``.
+
+    """
+
+    def __init__(self, storage: str = None, name: str = None, client: "distributed.Client" = None):
```

Comment:
storage should be `Optional[str, BaseStorage]`. name should be `Optional[str]` client should be `Optional["distributed.Client"]`.